### PR TITLE
PYTHON-2545 Test serverless proxy

### PR DESCRIPTION
**Pull request by prashantmital**
TO DO:

- [x] Update spec tests once https://github.com/mongodb/specifications/pull/1036 is merged
- [x] Add logic to skip non unified format tests that use features not supported by serverless (i.e. `collation` and `$out`)

### CODE REVIEW

File: test/__init__.py [hunk 1]
```diff
@@ -374,33 +388,36 @@ def _init_client(self):
...
-            self.server_parameters = self.client.admin.command(
-                'getParameter', '*')
 
-            if 'enableTestCommands=1' in self.cmd_line['argv']:
+            if TEST_SERVERLESS:
                 self.test_commands_enabled = True
-            elif 'parsed' in self.cmd_line:
-                params = self.cmd_line['parsed'].get('setParameter', [])
-                if 'enableTestCommands=1' in params:
+                self.has_ipv6 = False
```
Comment (ShaneHarvey) [inline test/__init__.py:394]: Serverless doesn't support ipv6?

File: .evergreen/config.yml [hunk 2]
```diff
@@ -836,9 +842,41 @@ post:
   - func: "cleanup"
   - func: "teardown_docker"
 
-tasks:
-
+task_groups:
```
[the remark this answers is not in the document]
Comment (prashantmital) [reply .evergreen/config.yml:845]: Yes. I too, haven't used them before. I picked this right out of the sample in https://github.com/mongodb-labs/drivers-evergreen-tools/blob/master/.evergreen/config.yml

File: .evergreen/config.yml [hunk 3]
```diff
@@ -836,9 +842,41 @@ post:
...
+      - command: shell.exec
+        params:
+          shell: "bash"
+          script: |
+            ${PREPARE_SHELL}
+            set +o xtrace
+            SERVERLESS_DRIVERS_GROUP=${SERVERLESS_DRIVERS_GROUP} \
+            SERVERLESS_API_PUBLIC_KEY=${SERVERLESS_API_PUBLIC_KEY} \
+            SERVERLESS_API_PRIVATE_KEY=${SERVERLESS_API_PRIVATE_KEY} \
+              bash ${DRIVERS_TOOLS}/.evergreen/serverless/create-instance.sh
```
[the remark this answers is not in the document]
Comment (ShaneHarvey) [reply .evergreen/config.yml:861]: Feel free to ignore this.

[the remark this answers is not in the document]
Comment (prashantmital) [reply .evergreen/config.yml:861]: Ignoring for now...

File: test/__init__.py (hunk 1 above)
Comment (prashantmital) [reply test/__init__.py:393]: I am not sure... It seems we only use this in one test (`TestClient.test_ipv6`) - do we want to run that test class against serverless?

Comment (ShaneHarvey) [reply test/__init__.py:393]: Ah okay, I don't think we need to run that test on serverless. A maybe useful test would be to manually resolve the serverless host to ipv6 and then attempt to connect. Up to you. I don't think it's needed. Should be something that's tested on the cloud team's side.

**Comment by ShaneHarvey** [comment]
Looks like there are still a large number of failures:
```
...
 [2021/07/09 20:16:42.584] FAIL [0.916s]: test_retryable_reads_listIndexes_serverErrors_ListIndexes_succeeds_after_SocketException (test_retryable_reads.TestSpec)
 [2021/07/09 20:16:42.584] ----------------------------------------------------------------------
 [2021/07/09 20:16:42.584] Traceback (most recent call last):
 [2021/07/09 20:16:42.584]   File "/data/mci/9b6fda5b61ca43196c9b54664e292149/src/test/__init__.py", line 534, in wrap
 [2021/07/09 20:16:42.584]     return f(*args, **kwargs)
 [2021/07/09 20:16:42.584]   File "/data/mci/9b6fda5b61ca43196c9b54664e292149/src/test/__init__.py", line 534, in wrap
 [2021/07/09 20:16:42.584]     return f(*args, **kwargs)
 [2021/07/09 20:16:42.584]   File "/data/mci/9b6fda5b61ca43196c9b54664e292149/src/test/test_retryable_reads.py", line 104, in run_scenario
 [2021/07/09 20:16:42.584]     self.run_scenario(scenario_def, test)
 [2021/07/09 20:16:42.584]   File "/data/mci/9b6fda5b61ca43196c9b54664e292149/src/test/utils_spec_runner.py", line 557, in run_scenario
 [2021/07/09 20:16:42.584]     self.check_events(test, listener, session_ids)
 [2021/07/09 20:16:42.584]   File "/data/mci/9b6fda5b61ca43196c9b54664e292149/src/test/utils_spec_runner.py", line 380, in check_events
 [2021/07/09 20:16:42.584]     len(res['started']), len(test['expectations']), cmds)
 [2021/07/09 20:16:42.584] AssertionError: 1 != 2 : [{'listIndexes': 'coll', 'cursor': {}, 'lsid': {'id': UUID('3045bd69-a342-4e1c-9442-ae30e0764147')}, '$clusterTime': {'clusterTime': Timestamp(1625861782, 4), 'signature': {'hash': b'\xbe[\xde\xe4%\xf8&\xd1\x84\xef_\xc1Y\x1c\xcc\x12\x0c0\xec\xe2', 'keyId': 6982211488521912322}}, '$db': 'retryable-reads-tests', '$readPreference': {'mode': 'primary'}}]
 [2021/07/09 20:16:42.584] ======================================================================
 [2021/07/09 20:16:42.584] FAIL [0.644s]: test_cluster_time_no_server_support (test_session.TestCausalConsistency)
 [2021/07/09 20:16:42.584] ----------------------------------------------------------------------
 [2021/07/09 20:16:42.584] Traceback (most recent call last):
 [2021/07/09 20:16:42.584]   File "/data/mci/9b6fda5b61ca43196c9b54664e292149/src/test/__init__.py", line 534, in wrap
 [2021/07/09 20:16:42.584]     return f(*args, **kwargs)
 [2021/07/09 20:16:42.584]   File "/data/mci/9b6fda5b61ca43196c9b54664e292149/src/test/test_session.py", line 1057, in test_cluster_time_no_server_support
 [2021/07/09 20:16:42.584]     self.assertIsNone(after_cluster_time)
 [2021/07/09 20:16:42.584] AssertionError: {'clusterTime': Timestamp(1625861787, 5), 'signature': {'hash': b"\xee\n<c\xa3\xce\xb0o\xf7\x18'P\x1c\x1d\x04\xe6\xcd\xb0\x1e\x8f", 'keyId': 6982211488521912322}} is not None
 [2021/07/09 20:16:42.584] ======================================================================
 [2021/07/09 20:16:42.584] FAIL [0.083s]: test_server_not_causal (test_session.TestCausalConsistency)
 [2021/07/09 20:16:42.584] ----------------------------------------------------------------------
 [2021/07/09 20:16:42.584] Traceback (most recent call last):
 [2021/07/09 20:16:42.584]   File "/data/mci/9b6fda5b61ca43196c9b54664e292149/src/test/__init__.py", line 534, in wrap
 [2021/07/09 20:16:42.584]     return f(*args, **kwargs)
 [2021/07/09 20:16:42.584]   File "/data/mci/9b6fda5b61ca43196c9b54664e292149/src/test/test_session.py", line 1016, in test_server_not_causal
 [2021/07/09 20:16:42.584]     self.assertIsNone(act)
 [2021/07/09 20:16:42.584] AssertionError: Timestamp(1625861787, 6) is not None
 [2021/07/09 20:16:42.690] ----------------------------------------------------------------------
 [2021/07/09 20:16:42.690] Ran 2042 tests in 426.268s
 [2021/07/09 20:16:42.690] FAILED (failures=198, errors=77, skipped=762)
 ```
 
 Waiting for those to be resolved before reviewing again.

File: test/test_crud_unified.py [hunk 4]
```diff
@@ -28,7 +28,8 @@
     os.path.dirname(os.path.realpath(__file__)), 'crud', 'unified')
 
 # Generate unified tests.
-globals().update(generate_test_classes(TEST_PATH, module=__name__))
+globals().update(generate_test_classes(
+    TEST_PATH, module=__name__, RUN_ON_SERVERLESS=True))
```
[the remark this answers is not in the document]
Comment (prashantmital) [reply test/test_crud_unified.py:32]: Not as per the wording of the spec. It only needs to be enabled for spec tests pertaining to certain driver specifications.

File: test/test_crud_v1.py [hunk 5]
```diff
@@ -131,8 +131,25 @@ def run_operation(collection, test):
...
+    if not client_context.serverless:
+        return
+
+    if test['operation']['name'] == 'aggregate':
+        for stage in test['operation']['arguments']['pipeline']:
+            if "$out" in stage:
+                self.skipTest("MongoDB Serverless does not support $out")
+
+    if "collation" in test['operation']['arguments']:
+        self.skipTest("MongoDB Serverless does not support collations")
```
[the remark this answers is not in the document]
Comment (prashantmital) [reply test/test_crud_v1.py:145]: The changes to make tests serverless friendly were only for unified format.

File: test/__init__.py [hunk 6]
```diff
@@ -334,10 +347,13 @@ def _init_client(self):
                 # May not have this if OperationFailure was raised earlier.
                 self.cmd_line = self.client.admin.command('getCmdLineOpts')
```
[the remark this answers is not in the document]
Comment (prashantmital) [reply test/__init__.py:348]: It works but doesn't give us any useful info:
```
self.client.admin.command('getCmdLineOpts')
{'argv': [], 'parsed': {}, 'ok': 1}
```

**Comment by prashantmital** [comment]
@ShaneHarvey this is ready for final review.

File: test/test_retryable_reads.py [hunk 7]
```diff
@@ -68,11 +69,25 @@ def maybe_skip_scenario(self, test):
...
+        if client_context.serverless:
+            for operation in test['operations']:
+                if operation['name'] == 'aggregate':
+                    for stage in operation['arguments']['pipeline']:
+                        if "$out" in stage:
+                            self.skipTest(
+                                "MongoDB Serverless does not support $out")
+                if "collation" in operation['arguments']:
+                    self.skipTest(
+                        "MongoDB Serverless does not support collations")
```
Comment (ShaneHarvey) [inline test/test_retryable_reads.py:82]: DRIVERS-1836 updated the unified tests and DRIVERS-1851 updated the crud v1 tests. Is there any reason why we did not update the retryable reads/writes tests? Can you file a drivers ticket for it?

**Review by ShaneHarvey — COMMENTED** [review]
Can you ensure the test suite fails in some way if the driver cannot connect to the serverless instance when `TEST_SERVERLESS==True`? For example you could add an extra test to TestClientContext. Or is this case already covered?

File: test/test_retryable_reads.py (hunk 7 above)
Comment (prashantmital) [reply test/test_retryable_reads.py:82]: I think this is only a concern for Retrayble Reads. I opened [DRIVERS-1869](https://jira.mongodb.org/browse/DRIVERS-1869).

File: test/test_client_context.py [hunk 8]
```diff
@@ -30,6 +30,15 @@ def test_must_connect(self):
                         'PYMONGO_MUST_CONNECT is set. Failed attempts:\n%s' %
                         (client_context.connection_attempt_info(),))
 
+    def test_serverless(self):
+        if 'TEST_SERVERLESS' not in os.environ:
+            raise SkipTest('TEST_SERVERLESS is not set')
+
+        self.assertTrue(client_context.connected,
```
Comment (ShaneHarvey) [inline test/test_client_context.py:37]: Suggest `client_context.connected and client_context.serverless`? Or is that redundant?

Comment (prashantmital) [reply test/test_client_context.py]: Somewhat redundant as `client_context.serverless` is set solely based on the value of `TEST_SERVERLESS` but it doesn't hurt. Updated.

**Review by ShaneHarvey — APPROVED** [review]

**Pull request merged**